### ISSUE TITLE
RequestIDs Repeating or Recurring over Short Periods

### ISSUE THREAD

**Issue by GLStephen** [prebid/Prebid.js#2822]
We are seeing instances of requestId (auctionId) recurring over short periods of time. Different IPs, pageviews and UAs. Is the requestId/auctionId generator sufficient to create true GUIDs? Just to be clear, we're 100% sure this is not some form of duplicate data in our ingestion.

**Comment by ScottKevill** [comment]
You're right. They rely on Math.random() which apparently has a long history of varying brokenness:

2011: [Google Chrome random number generator issue](http://devoluk.com/google-chrome-math-random-issue.html)
2011: [Collisions when generating UUIDs in JavaScript?](https://stackoverflow.com/questions/6906916/collisions-when-generating-uuids-in-javascript/)
2015: [TIFU by using Math.random()](https://medium.com/@betable/tifu-by-using-math-random-f1c308c4fd9d)
2015: [There's Math.random(), and then there's Math.random()](https://v8project.blogspot.com/2015/12/theres-mathrandom-and-then-theres.html)
2016: [Replace Math.random to prevent duplicate event IDs](https://github.com/snowplow/snowplow-javascript-tracker/issues/499)
2018: [Googlebot’s Javascript random() function is deterministic](http://www.tomanthony.co.uk/blog/googlebot-javascript-random/)


Here are the relevant Prebid functions:
https://github.com/prebid/Prebid.js/blob/a2bb2557a6cb45e91c6a5d38268ea257b9128ddc/src/utils.js#L46-L72

getUniqueIdentifierStr() is used for adId/bidId/requestId, and bidderRequestId.
generateUUID() is used for auctionId and transactionId.

Two suggested solutions are to either use [Crypto.getRandomValues](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues) where available, or instead include a Mersenne Twister implementation (eg. https://gist.github.com/banksean/300494).

**Comment by GLStephen** [comment]
@ScottKevill That is what I determined from reading the code and some of the same posts you mention above. Crypto.getRandomValues seems to have broad support. I'll look at patching it in for auctionId. Within the scope of an auction the current level of uniqueness appears to suffice.

**Comment by GLStephen** [comment]
@dbemiller do you have any insight on this? I'd like to make  PR for a change, but don't want to rehash old ground if there is existing discussion or work in the area which I could leverage.

**Comment by dbemiller** [comment]
I'm not aware of any previous discussions... but I probably wouldn't have been involved in the project that long ago.

I left my thoughts below:

---------------------------------------------

Your suggestions sound good to me. I personally think performance is overrated, and if we profiled this and found out that it was a bottleneck, it would be very easy to optimize (since it's limited to a single function & has no architectural impact). That said, your comment:

> Performance difference between the two appears to be sub millisecond on a typical browser.

is only fair if these functions are called a few times on a page. If you're worried about performance, it might be worth seeing how many calls there are when loading a typical page.

---------------------------------------------

I also skimmed the most popular [UUID-generating library](https://github.com/kelektiv/node-uuid) that I could find, and it [does use crypto](https://github.com/kelektiv/node-uuid/blob/master/lib/rng-browser.js) whenever it's available.

Also worth noting that, their `Math.random()` fallback makes 16 calls to `Math.random()`, whereas Prebid's only makes 1. From their comments:

> If all else fails, use Math.random().  It's fast, but is of unspecified quality.

so... this suggests two things:

1. It looks like 16 `Math.random()` calls are still faster than a single `crypto` one.
2. They still found that `Math.random()` had enough collisions that `crypto` was a better choice if available.

------------------------------------------------

I don't think the Mersenne twister is a good option here. If we generate a random seed, we would (probably?) have to use something like `Math.random()` to generate it--which puts us in the same boat. If we don't specify one, they use [the current time](https://gist.github.com/banksean/300494#file-mersenne-twister-js-L69). Since many users view your page simultaneously, I suspect high-traffic sites would see _more_ ID conflicts--not fewer.

**Comment by ScottKevill** [comment]
Not exactly, it's nuanced.

Even if Mersenne Twister were seeded with the current time, it would still be an improvement on `Math.random`, as `Math.random` has more problems than just its initial seed.

That said, the only reason to use Mersenne Twister would be if performance were a concern -- which I highly doubt, given how few random values are required with Prebid.

If enough random values *were* required that performance became a problem, then you'd seed Mersenne Twister with `crypto.getRandomValues` and only take the hit once. In this situation, if `crypto.getRandomValues` were not available, you could fall back to seeding with the current time and/or `Math.random`. The likelihood of collisions would still be lower than using time alone, because two users would need to be concurrent *and* be both missing browser support for `crypto.getRandomValues`.

**TL;DR:** Use `crypto.getRandomValues` and if anyone cares, `Math.random` where that is not available.

**Comment by dbemiller** [comment]
Agreed with all that ^^

**Comment by mkendall07** [comment]
hey all. Sorry was late to this thread. I'm in agreement with the suggested approach.
@GLStephen Feel free to open a PR. thanks!

**Comment by GLStephen** [comment]
@mkendall07 @dbemiller PR incoming after we finish internal testing, probably shouldn't be marked stale

**Comment by mkendall07** [comment]
@GLStephen any updates on this?

**Comment by GLStephen** [comment]
@mkendall07 PR is incoming
